File: C++/Hard/Accepted/maximal-rectangle_0.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        
        int m = matrix.size();
        if (m == 0) return 0;
        int n = matrix[0].size();
        if (n == 0) return 0;
        vector<int> left(n, 0), right(n, n), height(n, 0);
        int res = 0;
        for (int i = 0; i < m; i++) {
            int cur_left = 0, cur_right = n;
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    left[j] = max(left[j], cur_left);
                    height[j]++;
                } else {
                    left[j] = 0;
                    cur_left = j + 1;
                    height[j] = 0;
                }
            }
            for (int j = n - 1; j >= 0; j--) {
                if (matrix[i][j] == '1') {
                    right[j] = min(right[j], cur_right);
                } else {
                    right[j] = n;
                    cur_right = j;
                }
            }
            for (int j = 0; j < n; j++) {
                res = max(res, (right[j] - left[j]) * height[j]);
            }
        }
        return res;
    }
};

File: Java/Hard/Accepted/maximal-rectangle_0.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        
        int m = matrix.length;
        if (m == 0) {
            return 0;
        }
        int n = matrix[0].length;
        if (n == 0) {
            return 0;
        }
        int[] heights = new int[n];
        int[] lefts = new int[n];
        int[] rights = new int[n];
        Arrays.fill(rights, n);
        int max = 0;
        for (int i = 0; i < m; i++) {
            int left = 0;
            int right = n;
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    heights[j]++;
                    lefts[j] = Math.max(lefts[j], left);
                } else {
                    heights[j] = 0;
                    lefts[j] = 0;
                    left = j + 1;
                }
            }
            for (int j = n - 1; j >= 0; j--) {
                if (matrix[i][j] == '1') {
                    rights[j] = Math.min(rights[j], right);
                } else {
                    rights[j] = n;
                    right = j;
                }
            }
            for (int j = 0; j < n; j++) {
                max = Math.max(max, (rights[j] - lefts[j]) * heights[j]);
            }
        }
        return max;
    }
}